arch/x86/include/kernel_arch_func.h

```diff
@@ -19,7 +19,17 @@
 static inline bool arch_is_in_isr(void)
 {
 #ifdef CONFIG_SMP
-	return arch_curr_cpu()->nested != 0;
+	/* On SMP, there is a race vs. the current CPU changing if we
+	 * are preempted.  Need to mask interrupts while inspecting
+	 * (note deliberate lack of gcc size suffix on the
+	 * instructions, we need to work with both architectures here)
```

Comment:
what's the distinction between this technique and using arch_irq_lock()? is it for optimization?

Reply:
Purely optimization, yeah.  Actually I think there was a header ordering issue I was running into too, but this is definitely preferable because you don't need to inspect the RFLAGS value you pushed.

What would be even better would be to predicate this decision on something that isn't racy.  One idea that seems reasonable but I haven't tried is just testing RSP vs. the bounds of the interrupt stack(s) region.